### ISSUE TITLE
Implement needLocation logic in getListing()

### ISSUE THREAD

**Issue by shvachko** [GiraffaFS/giraffa#53]
Original [issue 53](https://code.google.com/p/giraffa/issues/detail?id=53) created by shvachko on 2012-10-14T22:54:29.000Z:

getListing() should not return locations if needLocation = false.

**Comment by shvachko** [comment]
Comment [#1](https://code.google.com/p/giraffa/issues/detail?id=53#c1) originally posted by shvachko on 2012-10-15T00:49:47.000Z:

Here is some work I did on this. It needs more checks around the creation of LocatedBlocks. Should we be implementing "getBlockLocations" here as well?

Let me know what you think.

**Comment by shvachko** [comment]
Comment [#2](https://code.google.com/p/giraffa/issues/detail?id=53#c2) originally posted by shvachko on 2012-10-15T02:58:27.000Z:

This looks good. Do you want to add a test case for that?

**Comment by shvachko** [comment]
Comment [#3](https://code.google.com/p/giraffa/issues/detail?id=53#c3) originally posted by shvachko on 2012-10-15T03:00:45.000Z:

Yeah I'll work on this after issue 50. Would you mind reviewing / +1'ing issue 33?

**Comment by shvachko** [comment]
Comment [#4](https://code.google.com/p/giraffa/issues/detail?id=53#c4) originally posted by shvachko on 2012-10-19T18:45:35.000Z:

The site was read-only this morning so I was not able to push out a patch. I had a question though: do we care about getListingInternal(....boolean needLocation)?

Basically what I am asking is -- do we want the logic to be that we ONLY load blocks (from HBase Result) if "needLocation" is true, or do we ALWAYS load blocks from Result and just create LocatedFileStatus (and pass the blocks) if "needLocation" is true? What is the right logic?

When do we need INode to know about its list of LocatedBlocks? Will it always be the case that ONLY if "needLocation" is true, only THEN we will need the blocks?

**Comment by shvachko** [comment]
Comment [#5](https://code.google.com/p/giraffa/issues/detail?id=53#c5) originally posted by shvachko on 2012-10-19T22:00:57.000Z:

That is a good question, which reminds me that we currently store blocks as a single row in NSTable. Ideally we should have two cells per block. One with block structure itself (id, gs, len) and second with that block locations. That way we can restrict the Get to not return columns with locations if needLocation = false. It will also let us access and update individual blocks when needed rather than the whole map as today. Let's deal with that redesign in the next issue.
For this issue let's just populate locations into LocatedFileStatus only if needLocation = true. Just a test case for now.

**Comment by shvachko** [comment]
Comment [#6](https://code.google.com/p/giraffa/issues/detail?id=53#c6) originally posted by shvachko on 2012-10-19T22:06:06.000Z:

Yeah I have a test case at home that is very simple right now. Should I add tests to TestGiraffaFS or make a new Test suite? EIther way is fine -- currently I made TestGiraffaFileStatus.

**Comment by shvachko** [comment]
Comment [#7](https://code.google.com/p/giraffa/issues/detail?id=53#c7) originally posted by shvachko on 2012-10-20T03:23:38.000Z:

This is my latest work. I struggled for a long time with how to incorporate "needLocation" and I believe it should be passed along to the getINode method because it is also used by getBlockLocations(), complete() and other methods.

I had to add the getClient() method in order to allow for the test to grab HdfsLocatedFileStatus. That method will need to exist for GiraffaAdmin anyway.

The test checks for the following:
needLocation = true on a completed File
needLocation = true on an incomplete File
needLocation = false on a completed File (assert instanceof HdfsFileStatus)

**Comment by shvachko** [comment]
Comment [#8](https://code.google.com/p/giraffa/issues/detail?id=53#c8) originally posted by shvachko on 2012-10-23T00:35:25.000Z:

In NamepsaceProcessor you can add getINode(src) that calls getINode(src, false), which the most common case. That way you will not need many changes that only add the parameter to getINode.
Not very good to expose the client only because it is needed in tests.
Can you instead make the test use HDFS or FileContext? This is where the functionality is actually utilized.

**Comment by shvachko** [comment]
Comment [#9](https://code.google.com/p/giraffa/issues/detail?id=53#c9) originally posted by shvachko on 2012-10-23T02:17:04.000Z:

Will do!

Not sure how to do it using HDFS -- we need to grab results from GiraffaClient. Perhaps FileContext will work.

**Comment by shvachko** [comment]
Comment [#10](https://code.google.com/p/giraffa/issues/detail?id=53#c10) originally posted by shvachko on 2012-10-23T04:10:00.000Z:

So your first suggestions were no problem, however, I could not get FileContext to work as it relies on using AbstractFileSystem which we need to either create one for GRFA. Another possible solution is to remove the "private" modifier on grfaClient and make it default package-protected instead. That way the test can use it... if you really don't want to expose the client AT ALL... we are gonna need to do something else. Here is my latest work.

It passes all current unit tests.

**Comment by shvachko** [comment]
Comment [#11](https://code.google.com/p/giraffa/issues/detail?id=53#c11) originally posted by shvachko on 2012-10-23T05:00:50.000Z:

You are right, FileContext is not straightforward. 
Let's make grfaClient package private as you propose.
Looks good. Let's ship it. +1

**Comment by shvachko** [comment]
Comment [#12](https://code.google.com/p/giraffa/issues/detail?id=53#c12) originally posted by shvachko on 2012-10-23T05:07:58.000Z:

Woot! Okay will do. Please check out issue 52, I think you'll like what I did there. :) (I hope) .. it is a major improvement, test-wise! We knocked out our testing time by 1/3!!

**Comment by shvachko** [comment]
Comment [#13](https://code.google.com/p/giraffa/issues/detail?id=53#c13) originally posted by shvachko on 2012-10-23T05:19:32.000Z:

I just committed this. I noticed an accident addition I had in GiraffaFileSystem.java and removed it -- I had a new constructor that was never being used. I had added it initially to try to use it for FileContext purposes but forgot to remove it before generating the patch.

Commit does not have this left-over constructor and I looked it over to make sure nothing else was committed but what was needed.

This is done.
